### PR TITLE
update opentelemetry to 0.20.0

### PR DESCRIPTION
**Pull request by ymgyt**
## Motivation

Update opentelemetry  dependency to use the latest opentelemetry from the tracing ecosystem.

## Solution

Modified code affected by opentelemetry changes

### CODE REVIEW

File: src/layer.rs [hunk 1]
```diff
@@ -1160,10 +1160,10 @@ mod tests {
     fn trace_id_from_existing_context() {
         let tracer = TestTracer(Arc::new(Mutex::new(None)));
         let subscriber = tracing_subscriber::registry().with(layer().with_tracer(tracer.clone()));
-        let trace_id = otel::TraceId::from(42u128.to_be_bytes());
+        let trace_id = otel::TraceId::from(42u128);
```
Comment (ymgyt) [inline src/layer.rs:1163]: use native from impl
https://github.com/open-telemetry/opentelemetry-rust/commit/782f9db07d1ed90451acf5b3521161e5dd1c8caf

File: src/lib.rs [hunk 2]
```diff
@@ -51,13 +51,17 @@
...
 //! use tracing_subscriber::layer::SubscriberExt;
 //! use tracing_subscriber::Registry;
 //!
-//! // Create a new OpenTelemetry pipeline
-//! let tracer = stdout::new_pipeline().install_simple();
+//! // Create a new OpenTelemetry trace pipeline that prints to stdout
+//! let provider = TracerProvider::builder()
+//!                .with_simple_exporter(opentelemetry_stdout::SpanExporter::default())
+//!                .build();
+//! let tracer = provider.tracer("readme_example");
```
Comment (ymgyt) [inline src/lib.rs:64]: I'm not sure this example is appropriate.
used the following
https://docs.rs/opentelemetry-stdout/latest/opentelemetry_stdout/index.html#examples

File: src/metrics.rs [hunk 3]
```diff
@@ -343,9 +326,16 @@ pub struct MetricsLayer {
 
 impl MetricsLayer {
     /// Create a new instance of MetricsLayer.
-    pub fn new(controller: BasicController) -> Self {
-        let meter =
-            controller.versioned_meter(INSTRUMENTATION_LIBRARY_NAME, Some(CARGO_PKG_VERSION), None);
+    pub fn new<M>(meter_provider: M) -> Self
+    where
+        M: MeterProvider,
```
Comment (ymgyt) [inline src/metrics.rs:331]: Changed dependency from sdk struct to api trait

File: src/metrics.rs [hunk 4]
```diff
@@ -343,9 +326,16 @@ pub struct MetricsLayer {
...
+    {
+        let meter = meter_provider.versioned_meter(
+            INSTRUMENTATION_LIBRARY_NAME,
+            Some(CARGO_PKG_VERSION),
+            None::<&'static str>,
```
Comment (ymgyt) [inline src/metrics.rs:336]: type annotation needed to compile

File: tests/metrics_publishing.rs [hunk 5]
```diff
@@ -156,126 +131,144 @@ async fn f64_histogram_is_exported() {
...
     tracing::subscriber::with_default(subscriber, || {
-        tracing::info!(value.abcdefg_racecar = 777.0012_f64);
+        tracing::info!(histogram.abcdefg_racecar = 777.0012_f64);
     });
 
     exporter.export().unwrap();
 }
 
-fn init_subscriber(
+fn init_subscriber<T>(
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:144]: sdk  removed `NumberKind`, so use generics

File: tests/metrics_publishing.rs [hunk 6]
```diff
@@ -156,126 +131,144 @@ async fn f64_histogram_is_exported() {
...
+
+    let provider = MeterProvider::builder().with_reader(reader.clone()).build();
     let exporter = TestExporter {
         expected_metric_name,
         expected_instrument_kind,
-        expected_number_kind,
         expected_value,
-        controller: controller.clone(),
+        reader,
+        _meter_provider: provider.clone(),
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:163]: We need hold `MeterProvider` until export complete.

`MeterProvider` hold `Pipeline` in `Arc`
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/metrics/meter_provider.rs#L28

`ManualReaderInner` hold reference to `Pipeline` in `Weak`
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/metrics/manual_reader.rs#L55

so, if not hold `MeterProvider`

```rust
tracing::subscriber::with_default(subscriber, || {
        tracing::info!(monotonic_counter.float_hello_world = 1.000000123_f64);
 });
// <- drop MeterProvider holded in MetricLayer

 exporter.export().unwrap(); // Weak<Pipeline>::upgrade() failed
```

File: tests/metrics_publishing.rs [hunk 7]
```diff
@@ -121,15 +101,11 @@ async fn f64_up_down_counter_is_exported() {
...
-        InstrumentKind::Histogram,
-        NumberKind::U64,
-        Number::from(9_u64),
-    );
+    let (subscriber, exporter) =
+        init_subscriber("abcdefg".to_string(), InstrumentKind::Histogram, 9_u64);
 
     tracing::subscriber::with_default(subscriber, || {
-        tracing::info!(value.abcdefg = 9_u64);
+        tracing::info!(histogram.abcdefg = 9_u64);
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:108]: previous histogram test span attribute prefix does not match impl (`histogram`)
https://github.com/tokio-rs/tracing-opentelemetry/blob/7b77d33d58eea0a7aa922bb45cfbab7272559c52/src/metrics.rs#L17

File: tests/metrics_publishing.rs [hunk 8]
```diff
@@ -156,126 +131,144 @@ async fn f64_histogram_is_exported() {
...
-                    self.expected_number_kind,
-                    *record.descriptor().number_kind()
-                );
+        let mut rm = ResourceMetrics {
+            resource: Resource::default(),
+            scope_metrics: Vec::new(),
+        };
+        self.reader.collect(&mut rm)?;
+
+        assert!(!rm.scope_metrics.is_empty());
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:236]: ensure at least one metrics exported

File: tests/metrics_publishing.rs [hunk 9]
```diff
@@ -156,126 +131,144 @@ async fn f64_histogram_is_exported() {
...
-                            assert_eq!(counts, &[0.0, 0.0, 1.0]);
-                        } else if self.expected_value.to_i64(&self.expected_number_kind) > 0 {
-                            assert_eq!(counts, &[0.0, 1.0, 0.0]);
-                        } else {
-                            assert_eq!(counts, &[1.0, 0.0, 0.0]);
-                        }
+                        let histogram =
+                            metric.data.as_any().downcast_ref::<Histogram<T>>().unwrap();
+                        let counts = histogram.data_points.first().unwrap().count;
+                        assert!(counts > 0);
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:263]: Do I need to be more strict with the histogram metrics assert?

File: src/lib.rs (hunk 2 above)
Comment (jtescher) [reply src/lib.rs:64]: this looks good, small nit that lines 62 and 63 are indented too many spaces

File: tests/metrics_publishing.rs (hunk 9 above)
Comment (jtescher) [reply tests/metrics_publishing.rs:263]: could add a sum assertion here too to ensure the value recorded is correct

File: tests/metrics_publishing.rs (hunk 6 above)
Comment (jtescher) [reply tests/metrics_publishing.rs:163]: Yep this looks right

**Review by jtescher — APPROVED** [review]
This looks great thanks @ymgyt, some small nits that can be done in a follow up if need be

File: src/lib.rs (hunk 2 above)
Comment (ymgyt) [reply src/lib.rs:64]: fixed indent

File: tests/metrics_publishing.rs [hunk 10]
```diff
@@ -156,126 +131,145 @@ async fn f64_histogram_is_exported() {
...
-                        } else if self.expected_value.to_i64(&self.expected_number_kind) > 0 {
-                            assert_eq!(counts, &[0.0, 1.0, 0.0]);
-                        } else {
-                            assert_eq!(counts, &[1.0, 0.0, 0.0]);
-                        }
+                        let histogram =
+                            metric.data.as_any().downcast_ref::<Histogram<T>>().unwrap();
+                        let histogram_data = histogram.data_points.first().unwrap();
+                        assert!(histogram_data.count > 0);
+                        assert_eq!(histogram_data.sum, self.expected_value);
```
Comment (ymgyt) [inline tests/metrics_publishing.rs:264]: add sum assertion

**Comment by ymgyt** [comment]
@jtescher Thanks for the review, I have fixed the indent in comment and added a sum assertion in histogram test .

**Comment by ymgyt** [comment]
CI is failing because pprof introduced in https://github.com/tokio-rs/tracing-opentelemetry/pull/27 is failing to build on windows.

https://github.com/tokio-rs/tracing-opentelemetry/actions/runs/5732443857/job/15537409319?pr=36

**Comment by jtescher** [comment]
yeah can fix that in a follow up 👍

**Pull request merged**